maro/simulator/scenarios/finance/business_engine.py

```diff
@@ -0,0 +1,653 @@
+
+    def _init_stocks(self):
+        # After frame initializing, it will help us create the station instances, let's create a reference.
+        # The attribute is added by frame that as same defined in frame definition.
+
+        # NOTE: Tthis is the build in station list that index start from 0,
+        # we need to create a mapping for it, as our trip data only contains id.
+        self._stocks = self._frame.stocks
+
+        for idx in range(len(self._conf["stocks"])):
+            # Get related station, and set the init states.
+            stock = self._stocks[idx]
+            stock.set_init_state(self._conf["stocks"][idx])
+
+    def _init_adj_matrix(self):
+        # # Our distance adj. Assume that the adj is NxN without header.
+        # distance_adj = np.array(load_adj_from_csv(self._conf["distance_adj_data"], skiprows=1))
+
+        # We only have one node here.
+        # self._matrices_node = self._frame.matrices[0]
+
+        # stock_num = len(self._stocks)
+
+        # self._distance_adj = distance_adj.reshape(station_num, station_num)
+
+        # Add wrapper to it to make it easy to use, with this we can get value by:
+        # 1. self._trips_adj[x, y].
+        # 2. self._trips_adj.get_row(0).
+        # 3. self._trips_adj.get_column(0).
+        # self._trips_adj = MatrixAttributeAccessor(self._matrices_node, "trips_adj", station_num, station_num)
+        pass
+
+    def _init_frame(self):
+        self._frame = build_frame(len(self._conf["stocks"]), self.calc_max_snapshots())
+        self._snapshots = self._frame.snapshots
+        self._account = self._frame.account[0]
+        self._account.set_init_state(
+            init_money=self._conf['account']['money']
+        )
+
+    def _register_events(self):
+        # Register our own events and their callback handlers.
+        self._event_buffer.register_event_handler(DECISION_EVENT, self._on_action_recieved)
+
+    def is_market_closed(self):
+        # TODO: Implement
+        return True
+
+    def _on_action_recieved(self, evt: Event):
+        actions = evt.payload
+        if actions is None:
+            return
+
+        for action in actions:
+            if action is None:
+                pass
+            else:
+                if isinstance(action, CancelOrder):
+                    self.cancel_order(action, evt.tick)
+                else:
+                    results: list = self.take_action(
+                        action, evt.tick
+                    )
+                    for result in results:
+                        self._account.take_trade(
+                            action, result, self._stocks
+                        )
+
+                if action.state != ActionState.pending and action.id not in self._finished_action:
+                    self._finished_action[action.id] = action
+        # append panding orders still not triggered in the available life time
+        if evt.tick in self._pending_orders:
```

Comment:
I don't understand the purpose of this part.

why use evt.tick here?

for tick in self._pending_orders:
    for action in self._pending_orders[tick]:
        ...

Reply:
self._pending_orders[tick] contains the not triggered actions that still live at the tick, fill them into event buffer at right time